File: 110_balanced_binary_tree/isBalanced2.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function isBalanced(root: TreeNode | null): boolean {
  if (root === null) return true;

  const left = getHeight(root.left);
  const right = getHeight(root.right);

  const skew = Math.abs(left - right);

  return skew <= 1 && isBalanced(root.left) && isBalanced(root.right);
}

function getHeight(root: TreeNode | null): number {
  if (root === null) return 0;

  const leftHeight = getHeight(root.left);
  const rightHeight = getHeight(root.right);

  return Math.max(leftHeight, rightHeight) + 1;
}

File: 110_balanced_binary_tree/isBalanced2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
ERR_CODE = -sys.maxsize - 1


class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        return self._get_height(root) != ERR_CODE

    def _get_height(self, root: TreeNode) -> int:
        if root:
            left = self._get_height(root.left)
            if left == ERR_CODE:
                return ERR_CODE
            right = self._get_height(root.right)
            if right == ERR_CODE:
                return ERR_CODE
            diff = abs(left - right)
            if diff > 1:
                return ERR_CODE
            return max(left, right) + 1

        return 0
